format/macho/macho.go

```diff
@@ -0,0 +1,868 @@
+			archBits = 32
+		} else {
+			archBits = 64
+		}
+	} else if magicBuffer == MH_CIGAM || magicBuffer == MH_CIGAM_64 {
+		d.Endian = decode.BigEndian
+		if magicBuffer == MH_CIGAM {
+			archBits = 32
+		} else {
+			archBits = 64
+		}
+	} else if magicBuffer == FAT_MAGIC {
+		d.Endian = decode.LittleEndian
+		fatParse(d)
+		return
+	} else if magicBuffer == FAT_CIGAM {
+		d.Endian = decode.BigEndian
+		fatParse(d)
+		return
+	} else {
+		// AR files are also valid OFiles but they should be parsed by `-d ar`
+		d.Fatalf("Invalid magic field")
+	}
+
+	d.SeekRel(-4 * 8)
+	d.FieldStruct("header", func(d *decode.D) {
+		d.FieldValueS("arch_bits", int64(archBits))
+		magic := d.FieldU32("magic", scalar.Hex, magicSymMapper)
+		d.FieldValueU("bits", uint64(archBits))
+		d.FieldValueStr("endian", endianNames[magic])
+		cpuType = d.FieldU32("cputype", cpuTypes)
+		d.FieldU32("cpusubtype", cpuSubTypes[cpuType])
+		d.FieldU32("filetype", fileTypes)
+		ncmds = d.FieldU32("ncdms")
+		d.FieldU32("sizeofncdms")
+		d.FieldStruct("flags", parseMachHeaderFlags)
+		if archBits == 64 {
+			d.FieldRawLen("reserved", 4*8, d.BitBufIsZero())
+		}
+	})
+	ncmdsIdx := 0
+	d.FieldStructArrayLoop("load_commands", "load_command", func() bool {
+		return ncmdsIdx < int(ncmds)
+	}, func(d *decode.D) {
+		cmd := d.FieldU32("cmd", loadCommands)
+		cmdsize := d.FieldU32("cmdsize")
+		switch cmd {
+		case LC_UUID:
+			d.FieldStruct("uuid_command", func(d *decode.D) {
+				d.FieldRawLen("uuid", 16*8)
+			})
+		case LC_SEGMENT, LC_SEGMENT_64:
+			// nsect := (cmdsize - uint64(archBits)) / uint64(archBits)
+			var nsects uint64
+			d.FieldStruct("segment_command", func(d *decode.D) {
+				d.FieldValueS("arch_bits", int64(archBits))
+				d.FieldUTF8NullFixedLen("segname", 16) // OPCODE_DECODER segname==__TEXT
+				if archBits == 32 {
+					d.FieldU32("vmaddr", scalar.Hex)
+					d.FieldU32("vmsize")
+					d.FieldU32("fileoff")
+					d.FieldU32("tfilesize")
+				} else {
+					d.FieldU64("vmaddr", scalar.Hex)
+					d.FieldU64("vmsize")
+					d.FieldU64("fileoff")
+					d.FieldU64("tfilesize")
+				}
+				d.FieldS32("initprot")
+				d.FieldS32("maxprot")
+				nsects = d.FieldU32("nsects")
+				d.FieldStruct("flags", parseSegmentFlags)
+			})
+			var nsectIdx uint64
+			d.FieldStructArrayLoop("sections", "section", func() bool {
+				return nsectIdx < nsects
+			},
+				func(d *decode.D) {
+					// OPCODE_DECODER sectname==__text
+					d.FieldUTF8NullFixedLen("sectname", 16)
+					d.FieldUTF8NullFixedLen("segname", 16)
+					if archBits == 32 {
+						d.FieldU32("address", scalar.Hex)
+						d.FieldU32("size")
+					} else {
+						d.FieldU64("address", scalar.Hex)
+						d.FieldU64("size")
+					}
+					d.FieldU32("offset")
+					d.FieldU32("align")
+					d.FieldU32("reloff")
+					d.FieldU32("nreloc")
+					// get section type
+					sectionType := d.U32()
+					sectionTypeMasked := sectionType & 0xff
+					d.FieldValueStr("type", sectionTypes[sectionTypeMasked])
```

Comment:
Thinking maybe `sectionType := d.U32() .... d. FieldValueStr()` can be replaced by something like:
```go
d.SeekRel(24)
d.FieldU8("type", sectionTypes)
d.SeekRel(-32)
d.FieldStruct("flags", parseSectionFlags)
...
```
or make `parseSectionFlags` only read 24 bits and have `d.FieldU8("type", sectionTypes)` after?

Is section type part of the flags in the spec?